pkg/graph/builder.go

```diff
@@ -682,117 +683,87 @@ func extractDependencies(env *cel.Env, expression string, resourceNames []string
+			}
 
-				_, err = dryRunExpression(env, expression, context)
-				if err != nil {
-					return fmt.Errorf("failed to dry-run expression %s: %w", expression, err)
-				}
+			// Type check the includeWhen expression
+			ast, iss := instanceEnv.Compile(includeWhenExpression)
+			if iss != nil && iss.Err() != nil {
+				return fmt.Errorf("failed to compile includeWhen expression %s: %w", includeWhenExpression, iss.Err())
 			}
 
-			// validate readyWhen Expressions for resource
-			// Only accepting expressions accessing the status and spec for now
-			// and need to evaluate to a boolean type
-			//
-			// TODO(michaelhtm) It shares some of the logic with the loop from above..maybe
-			// we can refactor them or put it in one function.
-			// I would also suggest separating the dryRuns of readyWhenExpressions
-			// and the resourceExpressions.
-			for _, readyWhenExpression := range resource.readyWhenExpressions {
-				fieldEnv, err := krocel.DefaultEnvironment(krocel.WithResourceIDs([]string{resource.id}))
-				if err != nil {
-					return fmt.Errorf("failed to create CEL environment: %w", err)
-				}
+			if ast.OutputType() != cel.BoolType {
+				return fmt.Errorf("includeWhen expression %s must return a boolean", includeWhenExpression)
+			}
+		}
 
-				err = validateCELExpressionContext(fieldEnv, readyWhenExpression, []string{resource.id})
-				if err != nil {
-					return fmt.Errorf("failed to validate expression context: '%s' %w", readyWhenExpression, err)
-				}
-				// create context
-				// add resource fields to the context
-				resourceEmulatedCopy := resource.emulatedObject.DeepCopy()
-				if resourceEmulatedCopy != nil && resourceEmulatedCopy.Object != nil {
-					delete(resourceEmulatedCopy.Object, "apiVersion")
-					delete(resourceEmulatedCopy.Object, "kind")
-				}
-				context := map[string]*Resource{}
-				context[resource.id] = &Resource{
-					emulatedObject: resourceEmulatedCopy,
-				}
-				output, err := dryRunExpression(fieldEnv, readyWhenExpression, context)
+		// For other expressions, we need to validate them conditionally
+		// based on the static analysis of includeWhen conditions
+		shouldValidateOtherExpressions := true
+		for _, includeWhenExpr := range resource.includeWhenExpressions {
+			// If the expression is a direct size comparison, we can statically analyze it
+			if strings.Contains(includeWhenExpr, "size()") && strings.Contains(includeWhenExpr, ">=") {
+				// Skip validation of other expressions if we can determine this resource won't be included
+				shouldValidateOtherExpressions = false
+				break
+			}
+		}
 
-				if err != nil {
-					return fmt.Errorf("failed to dry-run expression %s: %w", readyWhenExpression, err)
-				}
-				if !krocel.IsBoolType(output) {
-					return fmt.Errorf("output of readyWhen expression %s can only be of type bool", readyWhenExpression)
+		if shouldValidateOtherExpressions {
```

Comment:
I'm quiet sure understanding this part, i'm reading as, is a CEL expression does some sort of size or direct comparison do not validate the rest of the expressions. Am I getting it right?